### PR TITLE
feat(travis): Windows cross-compilation

### PR DESCRIPTION
**Pull request by nurupo**
~~~This is a Work In Progress (WIP) PR.~~~ It adds Windows cross-compilation build(s?) to Travis, since Windows builds frequently get broken without anyone realizing it. ~~~I will be messing with Travis and modifying scripts until I get Travis stages and the cross-compilation to work, so no point in reviewing before I'm done.~~~

**EDIT:** the PR is ready.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qtox/qtox/4736)
<!-- Reviewable:end -->

### CODE REVIEW

**Comment by Talkless** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[.travis/build-windows.sh, line 32 at r1](https://reviewable.io:443/reviews/qtox/qtox/4736#-KwQsOM3_mg3aIkwMCoX:-KwQsOM3_mg3aIkwMCoY:b55ko09) ([raw file](https://github.com/qtox/qtox/blob/944a71e4b16eb935595ba0d225046d28ed30cf5e/.travis/build-windows.sh#L32)):*
> ```Shell
> then
>   # Clear the cache, removing all the pre-built dependencies
>   rm -rf $DEP_CACHE/*
> ```

I'd suggest to quote every variable usage to avoid problems with possible spaces in paths.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4736)*

File: windows/cross-compile/build.sh [hunk 1]
```diff
@@ -211,30 +223,31 @@ then
...
   mkdir -p "$QT_PREFIX_DIR"
 
   QT_MIRROR=http://qt.mirror.constant.com
 
-  QT_MAJOR=5
-  QT_MINOR=6
-  QT_PATCH=3
-
-  QT_VERSION=$QT_MAJOR.$QT_MINOR.$QT_PATCH
   wget $QT_MIRROR/official_releases/qt/$QT_MAJOR.$QT_MINOR/$QT_VERSION/single/qt-everywhere-opensource-src-$QT_VERSION.tar.xz
```
Comment (sudden6) [inline windows/cross-compile/build.sh:272]: maybe add `--no-verbose` to produce less output?

Comment (nurupo) [reply windows/cross-compile/build.sh:272]: Nah, it's fine as it is, it doesn't hurt anything, not to mention that it's even beneficial to have a verbose output -- pretty sure whoever corss-compiles qTox on a slow internet would want to know the progress.

Comment (sudden6) [reply windows/cross-compile/build.sh:272]: this still displays errors, but hides the progress, which isn't that usefull imo

Comment (sudden6) [reply windows/cross-compile/build.sh:272]: e.g. in this job https://travis-ci.org/qTox/qTox/jobs/288231919 the download creates ~3k lines only for the progress :/

Comment (nurupo) [reply windows/cross-compile/build.sh:272]: This script is not only for Travis, but also for users and it's useful for them to see the progress. 3k for wget lines is a lot, I'm not sure why it does that on Travis, it doesn't print that many lines locally, but removing those ~3k lines won't help with anything, the entire build log is like 20k lines anyway, it doesn't even fit on Travis page, you have to access the log file directly. What would you gain by removing those 3k lines from Travis log and making users unable to track download progress?

**Comment by nurupo** [comment]
Weird, I can't reproduce any build failures I'm having on Jenkins. It builds just fine on Travis without https://github.com/qTox/qTox/commit/2f1bf107808953f5f080de3458f88abb33ad247a and https://github.com/qTox/qTox/commit/d69023c9d78714f5e730e98bb251c8d0f723688a commits.

I'm actually not sure about what is going on with `posixsignalnotifier` test at all. My PR branch doesn't have the https://github.com/qTox/qTox/commit/d69023c9d78714f5e730e98bb251c8d0f723688a commit yet, i.e. [there is no `if()` around `posixsignalnotifier` test in `cmake/Testing.cmake`](https://github.com/nurupo/qTox/blob/travis-windows/cmake/Testing.cmake), it should be build unconditionally with all other tests, yet `posixsignalnotifier` test is not even mentioned in build logs, the last test that is mentioned is `toxmedata`, I don't understand why it skips `posixsignalnotifier`.

**Comment by nurupo** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[.travis/build-windows.sh, line 32 at r1](https://reviewable.io:443/reviews/qtox/qtox/4736#-KwQsOM3_mg3aIkwMCoX:-Kwf_LX9e1X0Invv_HXh:b-c7n2mi) ([raw file](https://github.com/qtox/qtox/blob/944a71e4b16eb935595ba0d225046d28ed30cf5e/.travis/build-windows.sh#L32)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

I'd suggest to quote every variable usage to avoid problems with possible spaces in paths.
</blockquote></details>

Nah, it's fine. `CACHE_DIR` is set to either `/opt/build-windows/i686` or `/opt/build-windows/x86_64`, there are no spaces in them. There might be spaces in `$PWD`. so I quoted it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4736)*

**Comment by nurupo** [comment]
Should be good to go. A few things of note:

- Travis fails due to qTox build generating an warning and then treating it as an error.
- See https://github.com/qTox/qTox/pull/4736#issuecomment-336752910.
- You can notice that Windows jobs are inconsistent with all other jobs in `.travis.yml`. Windows jobs don't define `env: JOB=...` and they have their own `script:` and they pass arguments to that `script:` which otherwise be specified in `env:`. This inconsistency is required for Windows jobs to share the cache. [In Travis, jobs can share cache only if they use exactly the same `env:`, os, language, etc.](https://docs.travis-ci.com/user/caching/#Caches-and-build-matrices) I actually used this to force cache separation between `i686` and `x86_64` jobs by defining `env: CACHE_TRICK_VARIABLE` variae with value different for both; without this separation the caches of one arch were overwriting the cache of another arch.
- Cache is invalidated when `build.sh` is modified, which is the behavior you want. For example, if you open a PR that bumps versions in `build.sh`, you don't need to do any extra work to clear the cache, Travis will detect that `build.sh` was modified, wipe the cache and rebuild everything from scratch, caching the results.
- I have changed the distro from Ubuntu 16.04 to Debian Stretch because when I bumped Qt from 5.6.3 to 5.9.2 it was failing to build using Ubuntu 16.04's MinGW (5.3.1) but worked with Debian Stretch's MinGW (6.3.0).

**Comment by Talkless** [comment]
Review status: 0 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[.travis/build-windows.sh, line 32 at r1](https://reviewable.io:443/reviews/qtox/qtox/4736#-KwQsOM3_mg3aIkwMCoX:-Kwk5GwDaZIO8Nnxj5dD:b-f4k0zx) ([raw file](https://github.com/qtox/qtox/blob/944a71e4b16eb935595ba0d225046d28ed30cf5e/.travis/build-windows.sh#L32)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Nah, it's fine. `CACHE_DIR` is set to either `/opt/build-windows/i686` or `/opt/build-windows/x86_64`, there are no spaces in them. There might be spaces in `$PWD`. so I quoted it.
</blockquote></details>

I believe it would be _good practice_ to quote, well, everything. To get a "reflex" for any other future shell code. Deciding case-by-case basis might be too bug prone, and dangerous.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4736)*

**Comment by sudden6** [comment]
@nurupo can you check if  #4744 fixes the build error?

**Comment by nurupo** [comment]
@sudden6 sure, let me cherry-pick your commit into this PR and see what Travis says. Btw, you could as easily cherry-picked my commit into your PR and wouldn't have to wait on me to do so.

**Comment by nurupo** [comment]
Is anyone looking into https://github.com/qTox/qTox/pull/4736#issuecomment-336752910 ? The `posixsignalnotifier` test is a mystery to me.

**Comment by sudden6** [comment]
@nurupo sorry, didn't think to cherry-pick it into my PR...

About the posixsignalnotifier, I have no idea why it isn't compiled. On the other hand it doesn't really make sense to build any of the tests, since we can't run them anyway (without using things like wine).

**Comment by sudden6** [comment]
Reviewed 2 of 5 files at r3.
Review status: 2 of 5 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[.travis/build-windows.sh, line 32 at r1](https://reviewable.io:443/reviews/qtox/qtox/4736#-KwQsOM3_mg3aIkwMCoX:-KwoDfkrtKm4KmFdEME5:b5iruwu) ([raw file](https://github.com/qtox/qtox/blob/944a71e4b16eb935595ba0d225046d28ed30cf5e/.travis/build-windows.sh#L32)):*
<details><summary><i>Previously, Talkless (Vincas Dargis) wrote…</i></summary><blockquote>

I believe it would be _good practice_ to quote, well, everything. To get a "reflex" for any other future shell code. Deciding case-by-case basis might be too bug prone, and dangerous.
</blockquote></details>

I agree with @Talkless, we may as well do it right from the beginning

---

*[windows/cross-compile/build.sh, line 251 at r2](https://reviewable.io:443/reviews/qtox/qtox/4736#-KwPlUhvmKBK5vSjU-Pb-r2-251:-KwoOcxfqYLfAsUSJV4J:b-viwoid) ([raw file](https://github.com/qtox/qtox/blob/bdd43cb8c5a9075340e0bf56ea20bfe45e25c75e/windows/cross-compile/build.sh#L251)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

This script is not only for Travis, but also for users and it's useful for them to see the progress. 3k for wget lines is a lot, I'm not sure why it does that on Travis, it doesn't print that many lines locally, but removing those ~3k lines won't help with anything, the entire build log is like 20k lines anyway, it doesn't even fit on Travis page, you have to access the log file directly. What would you gain by removing those 3k lines from Travis log and making users unable to track download progress?
</blockquote></details>

Maybe only disable it on Travis then? I'll leave that up to you to decide though.

---

*[windows/cross-compile/build.sh, line 203 at r3](https://reviewable.io:443/reviews/qtox/qtox/4736#-KwoP26TsqU0e8IQ62Rg:-KwoP26TsqU0e8IQ62Rh:bq1gvxu) ([raw file](https://github.com/qtox/qtox/blob/5010c21823579de7aa44bcc77072d3ded0d089c5/windows/cross-compile/build.sh#L203)):*
> ```Shell
> 
>   wget https://www.openssl.org/source/openssl-$OPENSSL_VERSION.tar.gz
>   check_sha256 "ce07195b659e75f4e1db43552860070061f156a98bb37b672b101ba6e3ddf30c" "openssl-$OPENSSL_VERSION.tar.gz"
> ```

Maybe move the hash into a variable and put it to the others at the beginning? groups all variables needed for an update together.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4736)*

**Comment by sudden6** [comment]
The fix was merged, this should be rebased.

**Comment by nurupo** [comment]
Ok, I will rebase. I delayed the rebase in hope that someone would figure out why of all tests `posixsignalnotifier` test wasn't even being mentioned by cmake and make, since I didn't pull the commit that added the `if()` condition around it, i.e. all tests are unconditional

```CMake
auto_test(core toxpk)
auto_test(core toxid)
auto_test(chatlog textformatter)
auto_test(net toxmedata)
auto_test(platform posixsignalnotifier)
```

Oh well, I guess the mystery will go unsolved.

**Comment by nurupo** [comment]
Huh, weird. When I run the build in `debian:stretch-slim` container locally, it fails with

```
[ 40%] Building CXX object CMakeFiles/qtox_static.dir/src/net/toxmedata.cpp.obj
/build/qtox/src/net/toxmedata.cpp: In member function 'QString ToxmeData::createAddressRequest(ToxId, const QString&, const QString&, bool) const':
/build/qtox/src/net/toxmedata.cpp:123:63: error: 'time' was not declared in this scope
            "\",\"timestamp\":" + QString().setNum(time(nullptr)) +
                                                               ^
/build/qtox/src/net/toxmedata.cpp: In member function 'QString ToxmeData::deleteAddressRequest(const ToxPk&)':
/build/qtox/src/net/toxmedata.cpp:161:63: error: 'time' was not declared in this scope
            "\",\"timestamp\":" + QString().setNum(time(nullptr)) +
                                                               ^
CMakeFiles/qtox_static.dir/build.make:1267: recipe for target 'CMakeFiles/qtox_static.dir/src/net/toxmedata.cpp.obj' failed
make[2]: *** [CMakeFiles/qtox_static.dir/src/net/toxmedata.cpp.obj] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:219: recipe for target 'CMakeFiles/qtox_static.dir/all' failed
make[1]: *** [CMakeFiles/qtox_static.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
```

yet it doesn't on Travis. I'm confused as why. We use the same container that runs the same `build.sh` script. The only differences between my and Travis setup might be the Docker and Linux kernel versions.

**Comment by nurupo** [comment]
Haven't rebased yet, want to look a bit deeper into it. It's as if Travis runs this script on qTox master where all Windows errors are fixed instead of my `travis-windows` branch.

**Comment by nurupo** [comment]
I see. Turns out Travis merges my PR branch into master and then testes that. I always thought Travis was testing only my PR branch alone. In my PR branch there are code issues which should have failed the compilation, but they were fixed in master, and I couldn't figure out why in the world Travis wasn't failing. It acted as if it had the fixes from the master branch merged into my branch. Turns out this was indeed the case. Not sure if this is default Travis behavior or something that qTox repo has configured.

**Comment by nurupo** [comment]
Well, that solves the `posixsignalnotifier` test mystery. After the merge there was an `if(UNIX)` around it, so no wonder cmake and make never mentioned it.

**Comment by nurupo** [comment]
Review status: 2 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[.travis/build-windows.sh, line 32 at r1](https://reviewable.io:443/reviews/qtox/qtox/4736#-KwQsOM3_mg3aIkwMCoX:-Kwwq-XtIpGD3D-esVJO:b-11bl1i) ([raw file](https://github.com/qtox/qtox/blob/944a71e4b16eb935595ba0d225046d28ed30cf5e/.travis/build-windows.sh#L32)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

I agree with @Talkless, we may as well do it right from the beginning
</blockquote></details>

Ok, quoted every single variable, path or not path.

---

*[windows/cross-compile/build.sh, line 251 at r2](https://reviewable.io:443/reviews/qtox/qtox/4736#-KwPlUhvmKBK5vSjU-Pb-r2-251:-KwwqVv_7KrUT2Gjc4Rg:b1a18eu) ([raw file](https://github.com/qtox/qtox/blob/bdd43cb8c5a9075340e0bf56ea20bfe45e25c75e/windows/cross-compile/build.sh#L251)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

Maybe only disable it on Travis then? I'll leave that up to you to decide though.
</blockquote></details>

I think it's fine as it is.

---

*[windows/cross-compile/build.sh, line 203 at r3](https://reviewable.io:443/reviews/qtox/qtox/4736#-KwoP26TsqU0e8IQ62Rg:-Kwwq9HHfAymB663lmIy:bfwrr91) ([raw file](https://github.com/qtox/qtox/blob/5010c21823579de7aa44bcc77072d3ded0d089c5/windows/cross-compile/build.sh#L203)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

Maybe move the hash into a variable and put it to the others at the beginning? groups all variables needed for an update together.
</blockquote></details>

Done. Note that not all deps have hash. We get some of the deps through git, so there is really nothing to act as hash, aside from commit sha1, but we also use tag names in a few places.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4736)*

**Comment by sudden6** [comment]
:lgtm_strong: @Diadlo can you also take a look?

---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[windows/cross-compile/build.sh, line 203 at r3](https://reviewable.io:443/reviews/qtox/qtox/4736#-KwoP26TsqU0e8IQ62Rg:-Kwy53kNG2TgVSdNexXM:b-rbjuga) ([raw file](https://github.com/qtox/qtox/blob/5010c21823579de7aa44bcc77072d3ded0d089c5/windows/cross-compile/build.sh#L203)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

Done. Note that not all deps have hash. We get some of the deps through git, so there is really nothing to act as hash, aside from commit sha1, but we also use tag names in a few places.
</blockquote></details>

Ok this sounds good

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4736#-:-Kwy6fFA82b-ejOZFNAT:bbzeprc)*

**Comment by nurupo** [comment]
Btw, if I remember right, only OpenSSL and Qt5 provide sha256 hashes, I had to download all other tarballs and calculate it locally.

**Comment by Diadlo** [comment]
Reviewed 1 of 5 files at r3, 1 of 3 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[.travis.yml, line 32 at r4](https://reviewable.io:443/reviews/qtox/qtox/4736#-KwyXFKC749-hB-v6rPo:-KwyXFKC749-hB-v6rPp:b675wis) ([raw file](https://github.com/qtox/qtox/blob/062039f9e20afdb1d438448e366b65cb3f6f0bd3/.travis.yml#L32)):*
> ```YAML
>       os: linux
>       env: JOB=build-ubuntu-14-04
>     - stage: "Windows Stage 1: Dependencies (OpenSSL, Qt)"
> ```

What about adding architecure in stage name?
This will allow, when opening travis immediately understand which architecture have an error

---

*[windows/cross-compile/build.sh, line 190 at r5](https://reviewable.io:443/reviews/qtox/qtox/4736#-Kx1l14fgTbl7RT7kYpC:-Kx1l14fgTbl7RT7kYpD:b-yi3tif) ([raw file](https://github.com/qtox/qtox/blob/33b978fc31c784563187fb90853f6650ea268cb6/windows/cross-compile/build.sh#L190)):*
> ```Shell
>   fi
>   # Create a file listing hashes of all the files except .git/*
>   find . -type f | grep -v "^./.git" | LC_COLLATE=C sort --stable --ignore-case | xargs sha256sum > /tmp/hashes.sha
> ```

Why not check the commit hash?

---

*[windows/cross-compile/build.sh, line 218 at r5](https://reviewable.io:443/reviews/qtox/qtox/4736#-Kx1lT2PINqdgc0VF2yk:-Kx1lT2PINqdgc0VF2yl:b-9mzoox) ([raw file](https://github.com/qtox/qtox/blob/33b978fc31c784563187fb90853f6650ea268cb6/windows/cross-compile/build.sh#L218)):*
> ```Shell
>   mkdir -p "$OPENSSL_PREFIX_DIR"
> 
>   wget https://www.openssl.org/source/openssl-$OPENSSL_VERSION.tar.gz
> ```

What about creating function?
```
url=$1
hash=$2
wget $file
filename=`basename $url`
check_sha256 "$hash" "filename"

bsdtar -xf $filename
rm $filename
```

Or even whole "if [ -f done ] .... else fi" extract to a function with callbacks to build (not sure about this)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4736)*

**Comment by nurupo** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[.travis.yml, line 32 at r4](https://reviewable.io:443/reviews/qtox/qtox/4736#-KwyXFKC749-hB-v6rPo:-Kx1nSTleb4TaTalKP3U:bsnyj5z) ([raw file](https://github.com/qtox/qtox/blob/062039f9e20afdb1d438448e366b65cb3f6f0bd3/.travis.yml#L32)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

What about adding architecure in stage name?
This will allow, when opening travis immediately understand which architecture have an error
</blockquote></details>

Stages build all architectures though. Stage 1 builds both i686 and x86_64, same for stage 2 and stage 3.

---

*[windows/cross-compile/build.sh, line 190 at r5](https://reviewable.io:443/reviews/qtox/qtox/4736#-Kx1l14fgTbl7RT7kYpC:-Kx1nf7ZiDwIuD3qbSP_:b-v2h0e6) ([raw file](https://github.com/qtox/qtox/blob/33b978fc31c784563187fb90853f6650ea268cb6/windows/cross-compile/build.sh#L190)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Why not check the commit hash?
</blockquote></details>

Something something SHA1 vulnerable something.

---

*[windows/cross-compile/build.sh, line 218 at r5](https://reviewable.io:443/reviews/qtox/qtox/4736#-Kx1lT2PINqdgc0VF2yk:-Kx1npiKdC-lua3uJBYR:bx38pef) ([raw file](https://github.com/qtox/qtox/blob/33b978fc31c784563187fb90853f6650ea268cb6/windows/cross-compile/build.sh#L218)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

What about creating function?
```
url=$1
hash=$2
wget $file
filename=`basename $url`
check_sha256 "$hash" "filename"

bsdtar -xf $filename
rm $filename
```

Or even whole "if [ -f done ] .... else fi" extract to a function with callbacks to build (not sure about this)
</blockquote></details>

I was thinking about it and thought to leave it as it is, since some deps are wget'ed, some deps are git cloned and other were shallow git cloned on specific branch -- you'd have to write 3 separate functions for them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4736)*

**Comment by Diadlo** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[windows/cross-compile/build.sh, line 218 at r5](https://reviewable.io:443/reviews/qtox/qtox/4736#-Kx1lT2PINqdgc0VF2yk:-Kx1p7drGf7qvcE5pkz7:b-jrm956) ([raw file](https://github.com/qtox/qtox/blob/33b978fc31c784563187fb90853f6650ea268cb6/windows/cross-compile/build.sh#L218)):*
<details><summary><i>Previously, nurupo wrote…</i></summary><blockquote>

I was thinking about it and thought to leave it as it is, since some deps are wget'ed, some deps are git cloned and other were shallow git cloned on specific branch -- you'd have to write 3 separate functions for them.
</blockquote></details>

Here, at least 10 `wget` call :P
And just 3 git (which can be extracted in one function with not necessary branch parameter)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4736)*

**Comment by nurupo** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[windows/cross-compile/build.sh, line 218 at r5](https://reviewable.io:443/reviews/qtox/qtox/4736#-Kx1lT2PINqdgc0VF2yk:-Kx1plSRf57iTMoHiLDj:bpnzbsk) ([raw file](https://github.com/qtox/qtox/blob/33b978fc31c784563187fb90853f6650ea268cb6/windows/cross-compile/build.sh#L218)):*
<details><summary><i>Previously, Diadlo (Polshakov Dmitry) wrote…</i></summary><blockquote>

Here, at least 10 `wget` call :P
And just 3 git (which can be extracted in one function with not necessary branch parameter)
</blockquote></details>

Yeah, I replaced all shallow git clones with wgets with my last commit. Anyway, feel free to open a PR that extracts things in functions with callbacks, I'm not doing it in this PR though, I don't even know how.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4736)*

**Comment by nurupo** [comment]
Built all the 4 qTox builds and ran them on Windows, all seem to work fine, including gdb debugging.

**Comment by nurupo** [comment]
Btw, you can tell cmake to use wine for running tests by setting [`CMAKE_CROSSCOMPILING_EMULATOR` variable](https://cmake.org/cmake/help/v3.9/prop_tgt/CROSSCOMPILING_EMULATOR.html#prop_tgt:CROSSCOMPILING_EMULATOR) in the toolchain file. Want me to try setting this up?

**Comment by Diadlo** [comment]
:lgtm_strong:

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/qtox/qtox/4736#-:-Kx20Fgv2hPl-Im5oyFF:b-jbsf8z)*

**Comment by nurupo** [comment]
Oh, I forgot to remove some debug stuff, don't merge yet.

**Comment by nurupo** [comment]
Removed the stuff I used for debugging while the build script was failing.

**Comment by nurupo** [comment]
Also, I have tried running the tests in wine, and succeeded, but we can't do this the proper way by setting the [`CMAKE_CROSSCOMPILING_EMULATOR` variable](https://cmake.org/cmake/help/v3.9/prop_tgt/CROSSCOMPILING_EMULATOR.html#prop_tgt:CROSSCOMPILING_EMULATOR), because it breaks Qt5 cmake files as they use `add_custom_command()` internally and prefixing those with `wine` breaks them. So I have added just `CROSSCOMPILING_EMULATOR`, which is not some CMake's variable my mine (I could rename it if you don't want confusion) and prefixed the `add_test()`'s commands with it. Anyway, see the next commit I push.

**Comment by nurupo** [comment]
Also, the `${CHECK_LIBRARIES}` in `cmake/Testing.cmake` is bothering me. Where exactly does it come from? I suspect that it might always be empty. I haven't tried printing out contents of it yet, but I don't see what could be setting it, as there is no dependency on libcheck anywhere, nor it's set manually in some Travis bash script, like i do with `CROSSCOMPILING_EMULATOR`.

**Comment by Diadlo** [comment]
@nurupo Probably I forgot to remove `${CHECK_LIBRARIES}` when migrate to Qt tests. Feel free to remove

**Comment by Diadlo** [comment]
`CROSSCOMPILING_EMULATOR`. Wine is not emulator :smirk:
IMO better use `CROSSCOMPILING_STARTER` or `TEST_STARTER`

**Comment by nurupo** [comment]
>CROSSCOMPILING_EMULATOR. Wine is not emulator

Well, although Wine Is Not an Emulator, it's named after [`CMAKE_CROSSCOMPILING_EMULATOR`](https://cmake.org/cmake/help/v3.9/prop_tgt/CROSSCOMPILING_EMULATOR.html#prop_tgt:CROSSCOMPILING_EMULATOR). I guess I will rename it to `TEST_CROSSCOMPILING_EMULATOR` or something like this, since it's used only in `add_test()`.

**Comment by nurupo** [comment]
If you are building for arm on amd64, you could set it to an appropriate `qemu` command so that it would run your arm tests in an arm qemu vm, for example. If you feel adventurous enough, you could set it to a bash script which `scp`s the test binary to a remote server (i.e. raspberry pi) and then runs it through `ssh`. So it can be set to other things than just `wine`.

**Comment by nurupo** [comment]
Ok, changed `CROSSCOMPILING_EMULATOR` -> `TEST_CROSSCOMPILING_EMULATOR`.

**Comment by Diadlo** [comment]
@nurupo Is it ready to merge now?

**Comment by nurupo** [comment]
Yes, but let's first see if it builds successfully on Travis.

**Comment by nurupo** [comment]
If you want, we could also add `-gdwarf-2` to `CMAKE_CXX_FLAGS` for  `DEBUG and MINGW` configuration. `dwarf-2` is a debugging data format. This will allow wine's debugger to print a readable stack trace and show the line that resulted in a crash when running a test. Without setting this it will print some gibberish stack-trace and probably won't point to the correct line.

**Comment by nurupo** [comment]
Added `-gdwarf-2`. This results in the debug `qtox.exe` being 3mb larger (83mb vs 86mb), but should show more info if a test crashes in wine. This flag is needed primarily for wine, the gdb.exe that the debug qTox build includes can read the debug symbols without `-gdwarf-2` just fine.

**Comment by nurupo** [comment]
The PR is ready to be merged btw.

**Comment by Diadlo** [comment]
@sudden6 Still LGTM?

**Comment by sudden6** [comment]
@Diadlo yes

**Pull request merged**